sys/include/net/gnrc/dns/internals.h

```diff
@@ -0,0 +1,201 @@
+#include "net/gnrc/netreg.h"
+#include "net/gnrc/pktbuf.h"
+#include "net/gnrc/udp.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern kernel_pid_t _dns_pid;
+
+#if 1
+#define CT_HTONS(x)         ((		\
+		(x >> 8) & 0x00FF) | 		\
+		((x << 8) & 0xFF00))
+
+#define CT_HTONL(x)         ((		\
+		(x >> 24) & 0x000000FF) |	\
+		((x >> 8) & 0x0000FF00) |	\
+		((x << 8) & 0x00FF0000) |	\
+		((x << 24) & 0xFF000000))
+#else
+#define CT_HTONS(x)         (x)
+#define CT_HTONL(x)         (x)
+#endif
+
+typedef enum {
+    A                   = CT_HTONS(0x0001), /** IPv4 Query */
+    TXT                 = CT_HTONS(0x0010), /** Text Query */
+    AAAA                = CT_HTONS(0x001C), /** IPv6 Query */
+    SRV                 = CT_HTONS(0x0021), /** Service Query */
+} dns_query_type;
+
+typedef enum {
+    IN                  = CT_HTONS(0x0001), /** Service Query */
+} dns_query_class;
+
+typedef enum {
+    HDR_FLAGS_QR        = CT_HTONS(0x8000), /** Query (0), Response (1)*/
+    HDR_FLAGS_AA        = CT_HTONS(0x0400), /** Authoritative Answer. */
+    HDR_FLAGS_TC        = CT_HTONS(0x0200), /** TrunCation. */
+    HDR_FLAGS_RD        = CT_HTONS(0x0100), /** Recursion Desired. */
+    HDR_FLAGS_RA        = CT_HTONS(0x0080), /** Recursion Available. */
+} dns_hdr_flags_t;
+
+typedef enum {
+    GNRC_DNS_MSG_RESPONSE_TIMEDOUT          = 0x0300,
+    GNRC_DNS_MSG_RESOLVE_A                  = 0x0301,
+    GNRC_DNS_MSG_RESOLVE_AAAA               = 0x0302,
+    GNRC_DNS_MSG_RESOLVE_TXT                = 0x0303,
+    GNRC_DNS_MSG_RESOLVE_SRV                = 0x0304,
+    GNRC_DNS_MSG_SERVER_LIFETIME_EXPIRED    = 0x0305,
+    GNRC_DNS_MSG_QUERY_RESPONSE             = 0x0306,
+    GNRC_DNS_MSG_QUERY_TIMEDOUT             = 0x0307,
+
+    GNRC_DNS_MSG_MAX
+} gnrc_dns_msg;
+
+typedef struct __attribute__((packed)) {
+    uint16_t id;
+    uint16_t flags;
+    uint16_t questions;
+    uint16_t answers;
+    uint16_t authorities;
+    uint16_t additionals;
+} dns_header_t;
+
+typedef struct __attribute__((packed)) {
+    dns_query_type type         : 16;
+    dns_query_class class       : 16;
+} dns_query_footer_t;
+
+typedef struct __attribute__((packed)) {
+    network_uint16_t compressed;
+    union {
+        dns_query_type val      : 16;
+        network_uint16_t raw;
+    } type;union {
+        dns_query_class val     : 16;
+        network_uint16_t raw;
+    }class;
+    network_uint32_t ttl;
+    network_uint16_t data_len;
+    uint8_t data[];
+} dns_query_respones_t;
+
+/**
+ * @brief   DNS request information used during the request
+ */
+typedef struct {
+    uint32_t tries_left;
+    const char *query_str;
+    kernel_pid_t request_thread;
+    vtimer_t timeout_timer;
+    uint16_t request_id;
+    dns_query_type query_type;
+    gnrc_dns_response_t *response;
+} gnrc_dns_request_t;
+
+/**
+ * @brief   DNS server advertise information
+ */
+typedef struct {
+    mutex_t mutex;                                  /**< DNS mutex */
```

Comment:
what do you need this mutex for? You only have one thread, or do you intend to share your resource with other layers?